### PR TITLE
general: binary cache servers

### PR DESCRIPTION
**Pull request by tupakkatapa**
This pull request introduces setting up the build servers to [serve a Nix store via HTTP](https://nixos.org/manual/nix/stable/package-management/binary-cache-substituter.html#serving-a-nix-store-via-http) for our other machines. This means that the other machines can fetch already built packages from the build servers without needing to build them by themselves. This is necessary because building takes a lot of RAM, and it would be impractical to expect all machines to meet these high hardware specifications.

I have started this by setting up a `nix-serve` and an Nginx web server for the `hetzner-ephemeral-alpha` host by following [these](https://nixos.wiki/wiki/Binary_Cache) instructions provided by the NixOS wiki.

While researching how to do this, I came across [nix-serve-ng](https://github.com/aristanetworks/nix-serve-ng), which is allegedly faster and more reliable than the original `nix-serve`.  I suggest that we consider switching to it later if there are any issues with the original one.

Just for the record -- An alternative and authenticated way to accomplish this would be to provide SSH authentication for the client hosts to the build servers and allow them to be used as remote builders. This would also eliminate the need to ensure that the build server has the latest binaries built.

### CODE REVIEW

**Comment by tupakkatapa** [comment]
Debugging in progress, the current output:
```shell
[I] kari@torque ~ $ curl http://buidl0.ponkila.com/nix-cache-info
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

**Comment by tupakkatapa** [comment]
> Debugging in progress, the current output:
> 
> ```shell
> [I] kari@torque ~ $ curl http://buidl0.ponkila.com/nix-cache-info
> <html>
> <head><title>502 Bad Gateway</title></head>
> <body>
> <center><h1>502 Bad Gateway</h1></center>
> <hr><center>nginx</center>
> </body>
> </html>
> ```

For the record -- This also produced the following errors to the system logs:

```shell
Jul 21 20:40:44 hetzner-ephemeral-alpha (ve-start)[1461]: nix-serve.service: Failed to set up credentials: Protocol error
Jul 21 20:40:44 hetzner-ephemeral-alpha (ve-start)[1461]: nix-serve.service: Failed at step CREDENTIALS spawning /nix/store/grq0g4z12j0xcvpqy7axfyhivb5nd37r-unit-script-nix-serve-start/bin>
```
Removing `sops-nix` fixed this issue. I did not bother to find out the detailed reason why this was happening. I might try to debug this in the future if we decide to use sops for the build server hosts for some reason. 

If I had to guess, the issue is the exact one that this workaround resolves:

  https://github.com/ponkila/homestaking-infra/blob/9fb27d76551e1f4b348fb75763c9629923985f5a/hosts/ponkila-ephemeral-beta/default.nix#L11-L18

I even tried it with this commit: [`hetzner-ephemeral-alpha: possible nix-serve fix 5`](https://github.com/ponkila/homestaking-infra/pull/55/commits/0224b9f9c7c56ee3e7cf180087d8745467edcada), but couldn't get the system to kexec with it. This is probably just caused by something missing from the configuration that I overlooked.

**Comment by tupakkatapa** [comment]
To clarify, both build servers are currently using the better, [nix-serve-ng](https://github.com/aristanetworks/nix-serve-ng) version of the `nix-serve`.

---

Could not build and test the configuration for `hetzner-ephemeral-beta`, even though changes were made to address this issue at https://github.com/ponkila/homestaking-infra/issues/35. When compiling a Linux kernel with the host, it appears that the build fails, even with '--max-jobs 1'.

```shell
    ...
    For full logs, run 'nix log /nix/store/0sidy5cxbdkra444nlpjkgx7whd908mw-linux-6.1.38.drv'.
    note: build failure may have been caused by lack of free disk space
error: 1 dependencies of derivation '/nix/store/kxf1f83ikyqxvx3w9pxay8fbi6d6m65f-kexec-tree.drv' failed to build
```

File: .sops.yaml [hunk 1]
```diff
@@ -22,3 +24,9 @@ creation_rules:
         - *juuso
         age:
         - *dinar-ephemeral-beta
+  - path_regex: hosts/hetzner-ephemeral-alpha/secrets/[^/]+\.yaml$
+    key_groups:
+      - pgp:
```
[the remark this answers is not in the document]
Comment (tupakkatapa) [reply .sops.yaml]: This is not relevant anymore since I removed sops-nix for this host.

**Comment by jhvst** [comment]
Disk space running out is because our ARM build server only has 32GB of RAM, which is coincidentally the most that Hetzner provides for VMs. The better idea is to use the x86 server to QEMU bootstrap the ARM device. So, "wontfix" for now.

**Review by jhvst — APPROVED** [review]

**Comment by tupakkatapa** [comment]
LGTM, both servers are up and running with `nix-serve`. 

```shell
[I] kari@torque ~ $ curl http://buidl0.ponkila.com:5000/nix-cache-info
StoreDir: /nix/store
WantMassQuery: 1
Priority: 30
[I] kari@torque ~ $ curl http://buidl1.ponkila.com:5000/nix-cache-info
StoreDir: /nix/store
WantMassQuery: 1
Priority: 30
```
When building, fetching binaries seems to work as it should.

```shell
[I] kari@torque ~/W/p/homestaking-infra (feat/nix-serve) [1] $ nix build .#ponkila-ephemeral-beta --debug
...
finished download of 'http://buidl0.ponkila.com:5000/ah4rcsnd880xi5raf90rrnw5fi2xxcac.narinfo'; curl status = 0, HTTP status = 404, body = 14 bytes
checking substituter 'http://buidl1.ponkila.com:5000' for path '/nix/store/ah4rcsnd880xi5raf90rrnw5fi2xxcac-dirs-sys-0.3.7'
querying info about '/nix/store/ah4rcsnd880xi5raf90rrnw5fi2xxcac-dirs-sys-0.3.7' on 'http://buidl1.ponkila.com:5000'...
downloading 'http://buidl1.ponkila.com:5000/ah4rcsnd880xi5raf90rrnw5fi2xxcac.narinfo'...
starting download of http://buidl1.ponkila.com:5000/ah4rcsnd880xi5raf90rrnw5fi2xxcac.narinfo
finished download of 'https://cache.nixos.org/38y9v3wkc3l888yrprs7pmgqd9038ksr.narinfo'; curl status = 0, HTTP status = 200, body = 512 bytes
checking substituter 'http://buidl0.ponkila.com:5000' for path '/nix/store/2ilczynl95nzijgr8qbq9ld36xcyak0k-hyper-0.14.26'
querying info about '/nix/store/2ilczynl95nzijgr8qbq9ld36xcyak0k-hyper-0.14.26' on 'http://buidl0.ponkila.com:5000'...
downloading 'http://buidl0.ponkila.com:5000/2ilczynl95nzijgr8qbq9ld36xcyak0k.narinfo'...
starting download of http://buidl0.ponkila.com:5000/2ilczynl95nzijgr8qbq9ld36xcyak0k.narinfo
finished download of 'http://buidl1.ponkila.com:5000/gx4bacndxmydbjwmzgdvxq4v0y1cm9s7.narinfo'; curl status = 0, HTTP status = 404, body = 14 bytes
checking substituter 'https://cache.nixos.org' for path '/nix/store/gx4bacndxmydbjwmzgdvxq4v0y1cm9s7-core2-0.4.0'
querying info about '/nix/store/gx4bacndxmydbjwmzgdvxq4v0y1cm9s7-core2-0.4.0' on 'https://cache.nixos.org'...
downloading 'https://cache.nixos.org/gx4bacndxmydbjwmzgdvxq4v0y1cm9s7.narinfo'...
finished download of 'http://buidl1.ponkila.com:5000/hd9rjg8l7zn725l9h9sqq50lg7yzz2w1.narinfo'; curl status = 0, HTTP status = 404, body = 14 bytes
checking substituter 'https://cache.nixos.org' for path '/nix/store/hd9rjg8l7zn725l9h9sqq50lg7yzz2w1-base16ct-0.1.1'
starting download of https://cache.nixos.org/gx4bacndxmydbjwmzgdvxq4v0y1cm9s7.narinfo
querying info about '/nix/store/hd9rjg8l7zn725l9h9sqq50lg7yzz2w1-base16ct-0.1.1' on 'https://cache.nixos.org'...
downloading 'https://cache.nixos.org/hd9rjg8l7zn725l9h9sqq50lg7yzz2w1.narinfo'...
finished download of 'https://cache.nixos.org/bbxjcq9m9ah95mvsy6zmr82n20wwf59j.narinfo'; curl status = 0, HTTP status = 200, body = 528 bytes
checking substituter 'http://buidl0.ponkila.com:5000' for path '/nix/store/9mbaz8ijpyb5rc4fm94ybm4xj7v328ry-num_threads-0.1.6'
finished download of 'http://buidl0.ponkila.com:5000/239mj96m0243j66r0bacw10l6wvxs9v8.narinfo'; curl status = 0, HTTP status = 404, body = 14 bytes
...
```

**Comment by jhvst** [comment]
Nsq run needed

**Comment by tupakkatapa** [comment]
> Nsq run needed

Shoot.. already merged, pushed updated nix-store-queries to the main

**Pull request merged**